### DISCUSSION TITLE
Generate PDF from revealjs presentation (consistent with PDF Export in revealjs)

### DISCUSSION THREAD

**Discussion by jlperla** [quarto-dev/quarto-cli#7018] [Q&A] [answered]
### Description

I want to ship a pdf for my presentation, which is otherwise in revealjs.

Of course, I can try to use the `quarto render myslides.qmd --to pdf` etc. but I have very revealjs specific code in them, so the latex transformation fails.  Too much work to get them to both render.

But I notice that revealjs has a "pdf export" mode which works pretty well from a browser.  Is there a way that I can automate that output to genrate that view and then "print" it to a pdf direclty, emulating what a user would do if they did that?

**Comment by mcanouil** [comment]
You can use anything that can interact with a browser, such as <https://github.com/rstudio/webshot2> in R, etc.

For reference: <https://quarto.org/docs/presentations/revealjs/presenting.html#print-to-pdf>.

**Comment by jlperla** [comment]
Thanks.  SO there is no simple workflow to generate an HTML file then open in with some CLI tool on the print to pdf and print it out?  Might be a nice output type for quarto if there isn't?  I use python but am happy to install any pip packages/etc.

**Reply by mcanouil** [reply, marked as answer]
It depends what you mean by "simple", Reveal.js also advertise dectape for this, see <https://revealjs.com/pdf-export/>.

**Comment by jlperla** [comment]
Update: The `decktape` output largely worked but had a few quirks on installation (and font-embedding was confusing).

The `pdf` output sounds good in principle, but as soon as you go down the revealjs road you quickly add things incompatible with latex.

But it is pretty good.  I would suggest you consider a new output type `revealjs-pdf` or something like that which uses decktape.  I think it is a common enough feature that it would be worth adding at some point.

Wanted to reiterate what an amazing product quarto is.  It already has so many quality features, but I can now understand why you are doubling down on revealjs.

**Reply by mcanouil** [reply]
You can write this format or project type which uses decktape as a post-render script.
See the documentation about making extensions: <https://quarto.org/docs/extensions/>.

**Comment by jlperla** [comment]
Thanks.  I think the main benefit from having it builtin to a future quarto release for other people would be the installation integration.  It was a pain (and having an extension wouldn't help with that issue).

**Reply by mcanouil** [reply]
FYI, this is tracked at:
- https://github.com/quarto-dev/quarto-cli/issues/4677

Note that you can have a pre-render script in extension that perform the installation.

**Reply by agerlach** [reply]
I'm a little late to the party here, but just in case somebody else comes across this, the decktape docker images can be used to eliminate the pain of installing.

If you `cd` to your directory with your revealjs .html file you can run the following. Just change `PRESENTATIONNAME`

```
docker run --rm -t -v `pwd`:/slides -v $(pwd):/home/user astefanutti/decktape --screenshots-size 1920x1080 /home/user/PRESENTATIONNAME.html slides.pdf
```

Note that this maps your host working directory to /home/user in the container, so don't change the /home/user path for the html file.

**Comment by vqv** [comment]
I have run into this issue when creating PDF version of slides for a class so that my students can follow along and mark up on their iPads.   I could not get Decktape working, so I put together a simple solution using [Playwright](https://playwright.dev/).  I have a Python script that launches a headless Chromium browser, opens the Reveal JS presentation in print mode, and then prints to PDF.  It seems to work for my class slides and saves me the hassle of doing it manually with Chrome.

```python
import os
import argparse
from playwright.sync_api import sync_playwright

parser = argparse.ArgumentParser(description = "Generate PDF slides from Reveal.js HTML")
parser.add_argument("input", help = "HTML filename", type = os.path.abspath)
parser.add_argument("output", help = "PDF filename", type = os.path.abspath)
args = parser.parse_args()

with sync_playwright() as p:
  browser = p.chromium.launch()
  page = browser.new_page()
  url = "file://" + args.input

  page.goto(url + "?print-pdf", wait_until = "networkidle")
  page.locator(".reveal.ready").wait_for()
  page.pdf(path = args.output, prefer_css_page_size = True)
```

**Reply by cderv** [reply]
Thanks for sharing ! 

I added a note about this in
* https://github.com/quarto-dev/quarto-cli/issues/4677
